k8s/apis/solution/v1/instancehistory_webhook.go

```diff
@@ -0,0 +1,124 @@
+				Name:       instance.Name,
+				UID:        instance.UID,
+			}
+
+			if !configutils.CheckOwnerReferenceAlreadySet(r.OwnerReferences, ownerReference) {
+				r.OwnerReferences = append(r.OwnerReferences, ownerReference)
+			}
+			if r.Labels == nil {
+				r.Labels = make(map[string]string)
+			}
+			r.Labels[api_constants.RootResource] = r.Spec.RootResource
+		}
+	}
+
+	// Set annotation for the instance history
+	annotations := r.ObjectMeta.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	annotations = helper.AddAzureSystemDataAnnotations(annotations)
+	annotation_name := os.Getenv("ANNOTATION_KEY")
+	if annotation_name != "" {
+		parts := strings.Split(r.Name, constants.ResourceSeperator)
+		annotations[annotation_name] = parts[len(parts)-1]
+	}
+	r.ObjectMeta.SetAnnotations(annotations)
+}
+
+//+kubebuilder:webhook:path=/validate-solution-symphony-v1-instancehistory,mutating=false,failurePolicy=fail,sideEffects=None,groups=solution.symphony,resources=instancehistories,verbs=create,versions=v1,name=vinstancehistory.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &InstanceHistory{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *InstanceHistory) ValidateCreate() (admission.Warnings, error) {
+	resourceK8SId := r.GetNamespace() + "/" + r.GetName()
+	operationName := fmt.Sprintf("%s/%s", constants.InstanceHistoryOperationNamePrefix, constants.ActivityOperation_Write)
+	ctx := configutils.PopulateActivityAndDiagnosticsContextFromAnnotations(r.GetNamespace(), resourceK8SId, r.Annotations, operationName, historyReaderClient, context.TODO(), historyLog)
+
+	diagnostic.InfoWithCtx(historyLog, ctx, "validate create", "name", r.Name, "namespace", r.Namespace)
+	observ_utils.EmitUserAuditsLogs(ctx, "Instance history %s is being created on namespace %s", r.Name, r.Namespace)
+
+	return nil, nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *InstanceHistory) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	// TODO: instance history is readonly and should not be updated
+	return nil, nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *InstanceHistory) ValidateDelete() (admission.Warnings, error) {
+	resourceK8SId := r.GetNamespace() + "/" + r.GetName()
+	operationName := fmt.Sprintf("%s/%s", constants.InstanceHistoryOperationNamePrefix, constants.ActivityOperation_Delete)
+	ctx := configutils.PopulateActivityAndDiagnosticsContextFromAnnotations(r.GetNamespace(), resourceK8SId, r.Annotations, operationName, historyReaderClient, context.TODO(), historyLog)
+
+	diagnostic.InfoWithCtx(historyLog, ctx, "validate delete", "name", r.Name, "namespace", r.Namespace)
+	observ_utils.EmitUserAuditsLogs(ctx, "Instance history %s is being deleted on namespace %s", r.Name, r.Namespace)
+
+	return nil, nil
+}
```

Comment:
ValidateUpdate and ValidateDelete are not needed as you don't register Update and Delete checks. And do we need ValidateCreate?